src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -22,428 +22,698 @@ namespace NuGet.Commands
         {
-            var envVar = Environment.GetEnvironmentVariable("NUGET_XPROJ_WRITE_TARGETS");
+            // Generate file names
+            var targetsPath = string.Empty;
+            var propsPath = string.Empty;
 
-            bool forceWriteTargets = false;
-            if (!string.IsNullOrEmpty(envVar))
+            if (request.RestoreOutputType == RestoreOutputType.NETCore)
             {
-                Boolean.TryParse(envVar, out forceWriteTargets);
+                // PackageReference style projects
+                var projFileName = Path.GetFileName(request.Project.RestoreMetadata.ProjectPath);
+
+                targetsPath = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.targets");
+                propsPath = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.props");
             }
+            else
+            {
+                // Project.json style projects
+                var dir = Path.GetDirectoryName(project.FilePath);
 
-            return forceWriteTargets;
-        }
+                targetsPath = Path.Combine(dir, $"{project.Name}.nuget.targets");
+                propsPath = Path.Combine(dir, $"{project.Name}.nuget.props");
+            }
 
-        /// <summary>
-        /// Find all included msbuild assets for a graph.
-        /// </summary>
-        private static Dictionary<PackageIdentity, ContentItemGroup[]> GetMSBuildAssets(
-            RemoteWalkContext context,
-            RestoreTargetGraph graph,
-            PackageSpec project,
-            Dictionary<RestoreTargetGraph, Dictionary<string, LibraryIncludeFlags>> includeFlagGraphs,
-            PatternSet patternSet)
-        {
-            var buildGroupSets = new Dictionary<PackageIdentity, ContentItemGroup[]>();
+            // Targets files contain a macro for the repository root. If only the user package folder was used
+            // allow a replacement. If fallback folders were used the macro cannot be applied.
+            // Do not use macros for fallback folders. Use only the first repository which is the user folder.
+            var repositoryRoot = repositories.First().RepositoryRoot;
 
-            var flattenedFlags = IncludeFlagUtils.FlattenDependencyTypes(includeFlagGraphs, project, graph);
+            // Invalid msbuild projects should write out an msbuild error target
+            if (!targetGraphs.Any())
+            {
+                return GenerateMultiTargetFailureFiles(
+                    targetsPath,
+                    propsPath,
+                    request.RestoreOutputType,
+                    log,
+                    token);
+            }
 
-            // convert graph items to package dependency info list
-            var dependencies = ConvertToPackageDependencyInfo(graph.Flattened);
+            // Add additional conditionals for multi targeting
+            var multiTargetingFromMetadata = (request.Project.RestoreMetadata?.CrossTargeting == true);
```

Comment:
This is only internal so it seemed unneeded for now. There are many fields that still have the old project.json names such as SuppressParent that should be renamed to the new PackageReference equivalent. All of those changes for PackageSpec could be better done in a single change.